lm_eval/tasks/x_stance.py

```diff
@@ -0,0 +1,202 @@
+
+    def doc_to_text(self, doc):
+        return (
+            self.TOPIC
+            + doc["question"]
+            + "\n\n"
+            + self.OPINION
+            + doc["comment"]
+            + "\n\n"
+            + self.STANCE
+        )
+
+    def doc_to_target(self, doc):
+        target = doc["label"]
+        if target == 0:
+            return " " + self.AGAINST
+        else:
+            return " " + self.FAVOR
+
+    def construct_requests(self, doc, ctx):
+        """Uses RequestFactory to construct Requests and returns an iterable of
+        Requests which will be sent to the LM.
+
+        :param doc:
+            The document as returned from training_docs, validation_docs, or
+            test_docs.
+        :param ctx: str
+            The context string, generated by fewshot_context. This includes the natural
+            language description, as well as the few shot examples, and the question
+            part of the document for `doc`.
+        """
+        # rf.loglikelihood as the task is a classification problem. For each document the model predicts loglikelihood for the correct label
+        # ctx is the fully formatted fewshot example, i.e. K examples + comment to rate
+        ll_favor = rf.loglikelihood(ctx, " " + self.FAVOR)
+        ll_against = rf.loglikelihood(ctx, " " + self.AGAINST)
+
+        return ll_favor, ll_against
+
+    def process_results(self, doc, results):
+        """Take a single document and the LM results and evaluates, returning a
+        dict where keys are the names of submetrics and values are the values of
+        the metric for that one document
+
+        :param doc:
+            The document as returned from training_docs, validation_docs, or test_docs.
+        :param results:
+            The results of the requests created in construct_requests.
+        """
+
+        pred = 0
+        favor, against = results
+        # Evaluation metrics will only work with numerical labels
+        if favor[0] > against[0]:
+            pred = 1
+        else:
+            pred = 0
+        true_label = doc["label"]
+        return {
+            "acc": pred == true_label,
+            "precision": (true_label, pred),
+            "recall": (true_label, pred),
+            "f1": (true_label, pred),
+        }
+
+    def aggregation(self):
+        """
+        :returns: {str: [metric_score] -> float}
+            A dictionary where keys are the names of submetrics and values are
+            functions that aggregate a list of metric scores
+        """
+
+        return {
+            "acc": mean,
+            "precision": partial(_xstance_agg_precision, "precision"),
+            "recall": partial(_xstance_agg_recall, "recall"),
+            "f1": partial(_xstance_agg_f1, "f1"),
+        }
+
+    def higher_is_better(self):
+        return {"acc": True, "precision": True, "recall": True, "f1": True}
+
+
+# German part of the dataset
+class XStanceDE(XStance):
+    VERSION = 0
```

Comment:
VERSION + DATASET_PATH are not needed since they are not changed compared to the base class.